### DISCUSSION TITLE
Making a furuta pendulum with a stepper motor.

### DISCUSSION THREAD

**Discussion by DaxLynch** [gin66/FastAccelStepper#232] [General]
I am building a PID system to invert a pendulum, similar to shown the one shown in this video [these](https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DXKzzWe15DEw&psig=AOvVaw3QBYteTg2fg3pJ9O0DIArw&ust=1707034582260000&source=images&cd=vfe&opi=89978449&ved=2ahUKEwii7rbA3Y6EAxX1HDQIHe3ZDEkQr4kDegQIARBv). I am using the FastAccel library, but I am struggling to figure out how structure my acceleration commands. 

What I want is if the PID code decides to make an action, i.e. provide an impulse, than I would like the stepper to accelerate in the desired direction, at possibly a varied acceleration rate. The problem I was thinking would occur is that these PID measurement and reaction cycles take place about every 10ms, so if I use moveByAcceleration, the stepper queue is still likely to be working on the previous commands, and if the PID command dictates that a direction command needs to take place, than the queue will have to go through all previous acceleration commands to get there.

I think what I am wondering is there a method through this library that I can have the stepper accelerating at a rate, and then continue at that rate, or change either the rate or direction, at a rate of once every 10ms?

**Comment by gin66** [comment]
The queue only contains stepper commands (generate steps at a given rate or execute a pause) and no acceleration value changes. The StepperTask is filling the queue, so that at least `20ms = 1s/50` in the future is covered in the queue.

```
while (!isQueueFull() &&
         ((ticksPrepared < TICKS_PER_S / 50) || q->queueEntries() <= 1) &&
         _rg.isRampGeneratorActive()) {
```

So there could be these two approaches:

1. In first step, patch the value of 50 to e.g. 200, so that the preplanning is only 5ms into the future. It should not be less than 5ms, because the stepper task only runs every 4ms. If that works, perhaps the library can be extended with another API call to set this value.
2. Most likely the acceleration is not needed at all and your code fills the queue manually using the [low level command API](https://github.com/gin66/FastAccelStepper/blob/master/extras/doc/FastAccelStepper_API.md#low-level-stepper-queue-management-low-level-access)

**Reply by DaxLynch** [reply]
Thank you for this repsonse and library. I have made a lot of progress on my project. Currently I am confused about how the step commands work. What do the ticks and steps refer to? Does it tell the stepper to do x steps in y ticks, at the rate of my DUE's TICKS_PER_S?

**Reply by gin66** [reply]
Just see, that the stepper command is not explained. Perhaps this helps (https://github.com/gin66/FastAccelStepper/issues/86)

**Reply by DaxLynch** [reply]
I want to use timer based interrupts and only 1 stepper motor, is there a library or method you think I should do? Should I do it just from the PWM channel? Is there a compatible library?

**Reply by gin66** [reply]
not clear, what you are looking for
